cpp/src/join/joining.cu

```diff
@@ -242,149 +210,175 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
-        CUDA_TRY( cudaMemset(result_cols[i]->valid, 0, sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size)) );
-        CHECK_STREAM(0);
+    
+    // Accumalate the right non-join col
+    for (int rindex = 0; rindex < num_right_cols-num_merging_join_cols; ++rindex)
+    {
+        result_rnonjoincol.push_back(result_right.get_column(rindex));
     }
-
-
-    // If the join_type is an outer join, then indices for non-matches will be
-    // -1, requiring bounds checking when gathering the result table
+ 
     bool const check_bounds{ join_type != JoinType::INNER_JOIN };
 
     // Construct the left columns
     if (0 != lnonjoincol.size()) {
-      cudf::table left_source_table(lnonjoincol.data(), lnonjoincol.size());
-      cudf::table left_destination_table(result_cols,
-                                         num_left_cols - num_cols_to_join);
+      cudf::table left_source_table(lnonjoincol);
+      cudf::table left_destination_table(result_lnonjoincol);
 
       cudf::detail::gather(&left_source_table,
                            static_cast<index_type const *>(left_indices->data),
                            &left_destination_table, check_bounds);
+
       CHECK_STREAM(0);
       gdf_error update_err = nvcategory_gather_table(left_source_table,left_destination_table);
       CHECK_STREAM(0);
-      GDF_REQUIRE(update_err == GDF_SUCCESS,update_err);
+      CUDF_EXPECTS(update_err == GDF_SUCCESS, "nvcategory_gather_table throwing a GDF error");
     }
-
+    
     // Construct the right columns
     if (0 != rnonjoincol.size()) {
-      cudf::table right_source_table(rnonjoincol.data(), rnonjoincol.size());
-      cudf::table right_destination_table(result_cols + right_table_begin,
-                                          num_right_cols - num_cols_to_join);
+      cudf::table right_source_table(rnonjoincol);
+      cudf::table right_destination_table(result_rnonjoincol);
 
       cudf::detail::gather(&right_source_table,
                            static_cast<index_type const *>(right_indices->data),
                            &right_destination_table, check_bounds);
       CHECK_STREAM(0);
       gdf_error update_err = nvcategory_gather_table(right_source_table,right_destination_table);
       CHECK_STREAM(0);
-      GDF_REQUIRE(update_err == GDF_SUCCESS,update_err);
+      CUDF_EXPECTS(update_err == GDF_SUCCESS, "nvcategory_gather_table throwing a GDF error");
     }
 
     // Construct the joined columns
-    if (0 != ljoincol.size()) {
-      cudf::table join_source_table(ljoincol.data(), ljoincol.size());
-      cudf::table join_destination_table(result_cols + left_table_end,
-                                         num_cols_to_join);
+    if (0 != ljoincols.num_columns() && num_merging_join_cols > 0) {
+
+      std::vector <gdf_column *> l_join;
+      std::vector <gdf_column *> r_join;
+      for (int join_ind = 0; join_ind < num_merging_join_cols; ++join_ind)
```

Comment:
I think this is a `set_intersection` followed by a gather...

Reply:
The indices of left and right indices are kind of pairs and need not be in order, so set_intersection will not work.